### ISSUE TITLE
FTL Crash on version 5.1

### ISSUE THREAD

**Issue by clarkchentw** [pi-hole/FTL#842]
```
[2020-07-19 05:43:09.244 138381/T138385] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[2020-07-19 05:43:09.244 138381/T138385] ---------------------------->  FTL crashed!  <----------------------------
[2020-07-19 05:43:09.244 138381/T138385] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[2020-07-19 05:43:09.244 138381/T138385] Please report a bug at https://github.com/pi-hole/FTL/issues
[2020-07-19 05:43:09.244 138381/T138385] and include in your report already the following details:
[2020-07-19 05:43:09.244 138381/T138385] FTL has been running for 56 seconds
[2020-07-19 05:43:09.244 138381/T138385] FTL branch: master
[2020-07-19 05:43:09.244 138381/T138385] FTL version: v5.1
[2020-07-19 05:43:09.244 138381/T138385] FTL commit: b522646
[2020-07-19 05:43:09.244 138381/T138385] FTL date: 2020-07-15 22:26:33 +0100
[2020-07-19 05:43:09.244 138381/T138385] FTL user: started as pihole, ended as pihole
[2020-07-19 05:43:09.244 138381/T138385] Compiled for x86_64 (compiled on CI) using gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
[2020-07-19 05:43:09.244 138381/T138385] Process details: MID: 138381
[2020-07-19 05:43:09.244 138381/T138385]                  PID: 138381
[2020-07-19 05:43:09.244 138381/T138385]                  TID: 138385
[2020-07-19 05:43:09.244 138381/T138385]                  Name: database
[2020-07-19 05:43:09.244 138381/T138385] Received signal: Segmentation fault
[2020-07-19 05:43:09.244 138381/T138385]      at address: 0x28
[2020-07-19 05:43:09.244 138381/T138385]      with code:  SEGV_MAPERR (Address not mapped to object)
```

**Comment by DL6ER** [comment]
Can you please post some more lines above the initial `!!!` ?

**Comment by clarkchentw** [comment]
2020-07-19 05:42:18.136 138377M] Resizing "/FTL-queries" from 73924608 to 74186752
[2020-07-19 05:42:18.143 138377M] Resizing "/FTL-queries" from 74186752 to 74448896
[2020-07-19 05:42:18.150 138377M] Resizing "/FTL-queries" from 74448896 to 74711040
[2020-07-19 05:42:18.157 138377M] Resizing "/FTL-queries" from 74711040 to 74973184
[2020-07-19 05:42:18.161 138377M] Imported 1169804 queries from the long-term database
[2020-07-19 05:42:18.162 138377M]  -> Total DNS queries: 1169804
[2020-07-19 05:42:18.162 138377M]  -> Cached DNS queries: 1122507
[2020-07-19 05:42:18.162 138377M]  -> Forwarded DNS queries: 47297
[2020-07-19 05:42:18.162 138377M]  -> Blocked DNS queries: 0
[2020-07-19 05:42:18.162 138377M]  -> Unknown DNS queries: 0
[2020-07-19 05:42:18.162 138377M]  -> Unique domains: 573
[2020-07-19 05:42:18.162 138377M]  -> Unique clients: 8
[2020-07-19 05:42:18.162 138377M]  -> Known forward destinations: 2
[2020-07-19 05:42:18.162 138377M] Successfully accessed setupVars.conf
[2020-07-19 05:42:19.463 138381M] PID of FTL process: 138381
[2020-07-19 05:42:19.463 138381M] Reloading DNS cache
[2020-07-19 05:42:19.463 138381M] Blocking status is enabled
[2020-07-19 05:42:19.464 138381/T138384] Listening on Unix socket
[2020-07-19 05:42:19.464 138381/T138383] Listening on port 4711 for incoming IPv6 telnet connectio
ns
[2020-07-19 05:42:19.464 138381/T138382] Listening on port 4711 for incoming IPv4 telnet connectio
ns
[2020-07-19 05:42:22.074 138381M] INFO: No regex blacklist entries found
[2020-07-19 05:42:22.074 138381M] INFO: No regex whitelist entries found
[2020-07-19 05:42:22.075 138381M] Compiled 0 whitelist and 0 blacklist regex filters for 8 clients
 in 1.4 msec
[2020-07-19 05:43:07.709 138381M] Shutting down...
[2020-07-19 05:43:09.240 138381M] Finished final database update

**Comment by clarkchentw** [comment]
[2020-07-19 05:43:09.244 138381/T138385]      at address: 0x28
[2020-07-19 05:43:09.244 138381/T138385]      with code:  SEGV_MAPERR (Address not mapped to objec
t)
[2020-07-19 05:43:09.250 138381M] ########## FTL terminated after 5.583105e+01 s! ##########
[2020-07-19 05:43:11.009 138510M] Using log file /var/log/pihole-FTL.log
[2020-07-19 05:43:11.009 138510M] ########## FTL started! ##########
[2020-07-19 05:43:11.009 138510M] FTL branch: master
[2020-07-19 05:43:11.009 138510M] FTL version: v5.1
[2020-07-19 05:43:11.010 138510M] FTL commit: b522646
[2020-07-19 05:43:11.010 138510M] FTL date: 2020-07-15 22:26:33 +0100
[2020-07-19 05:43:11.010 138510M] FTL user: pihole
[2020-07-19 05:43:11.010 138510M] Compiled for x86_64 (compiled on CI) using gcc (Debian 6.3.0-18+
deb9u1) 6.3.0 20170516
[2020-07-19 05:43:11.010 138510M] Starting config file parsing (/etc/pihole/pihole-FTL.conf)
[2020-07-19 05:43:11.010 138510M]    SOCKET_LISTENING: only local
[2020-07-19 05:43:11.010 138510M]    AAAA_QUERY_ANALYSIS: Show AAAA queries
[2020-07-19 05:43:11.010 138510M]    MAXDBDAYS: max age for stored queries is 365 days
[2020-07-19 05:43:11.010 138510M]    RESOLVE_IPV6: Resolve IPv6 addresses
[2020-07-19 05:43:11.010 138510M]    RESOLVE_IPV4: Resolve IPv4 addresses
[2020-07-19 05:43:11.010 138510M]    DBINTERVAL: saving to DB file every minute
[2020-07-19 05:43:11.010 138510M]    DBFILE: Using /etc/pihole/pihole-FTL.db
[2020-07-19 05:43:11.010 138510M]    MAXLOGAGE: Importing up to 24.0 hours of log data
[2020-07-19 05:43:11.010 138510M]    PRIVACYLEVEL: Set to 0
[2020-07-19 05:43:11.010 138510M]    IGNORE_LOCALHOST: Show queries from localhost
[2020-07-19 05:43:11.010 138510M]    BLOCKINGMODE: Null IPs for blocked domains
[2020-07-19 05:43:11.010 138510M]    ANALYZE_ONLY_A_AND_AAAA: Disabled. Analyzing all queries
[2020-07-19 05:43:11.010 138510M]    DBIMPORT: Importing history from database
[2020-07-19 05:43:11.010 138510M]    PIDFILE: Using /run/pihole-FTL.pid
[2020-07-19 05:43:11.010 138510M]    SOCKETFILE: Using /run/pihole/FTL.sock

**Comment by clarkchentw** [comment]
Just FYI, I have a cronjob automatically detects whether port 53 is listening or not, if not it restarts. So you might see it restart after it crash within a minutes.

**Comment by DL6ER** [comment]
> I have a cronjob automatically detects whether port 53 is listening or not, if not it restarts

Hmm, yeah, this explains why we see:
```
[2020-07-19 05:43:09.244 138381/T138385]      with code:  SEGV_MAPERR (Address not mapped to object)
<->
[2020-07-19 05:43:09.250 138381M] ########## FTL terminated after 5.583105e+01 s! ##########
[2020-07-19 05:43:11.009 138510M] Using log file /var/log/pihole-FTL.log
[2020-07-19 05:43:11.009 138510M] ########## FTL started! ##########
```

Where I put `<->`, I'd have expected to see a backtrace, however, your cronjob apparently terminated FTL before it was able to generate the backtrace. Can you afford running the cron job maybe only every five minutes (like `*/5 * * * *`) so we have a chance to pick up the backtrace? It will tell us *where* the crash happened inside FTL.

**Comment by DL6ER** [comment]
Also, please try if
```
pihole checkout ftl fix/guard_gravity_reopening
```
fixes the issues you're seeing.

**Comment by clarkchentw** [comment]
Fix applied, also I added a 60 seconds delay before calling restart. Let's see how it goes.

**Comment by g0rbe** [comment]
```
Resizing "/FTL-dns-cache" from 2125824 to 2129920
[2020-07-25 00:01:10.005 9057/T22086] Remapping "/FTL-dns-cache" from 2125824 to 2129920
[2020-07-25 00:14:30.487 26482/F9057] Resizing "/FTL-strings" from 3219456 to 3223552
[2020-07-25 00:14:30.490 9057M] Remapping "/FTL-strings" from 3219456 to 3223552
[2020-07-25 00:22:21.750 29022/F9057] Resizing "/FTL-dns-cache" from 2129920 to 2134016
[2020-07-25 00:22:22.004 9057/T29023] Remapping "/FTL-dns-cache" from 2129920 to 2134016
[2020-07-25 00:33:06.904 32487/F9057] Resizing "/FTL-strings" from 3223552 to 3227648
[2020-07-25 00:33:06.005 9057/T32489] Remapping "/FTL-strings" from 3223552 to 3227648
[2020-07-25 00:46:44.837 4477/F9057] Resizing "/FTL-dns-cache" from 2134016 to 2138112
[2020-07-25 00:46:45.004 9057/T4485] Remapping "/FTL-dns-cache" from 2134016 to 2138112
[2020-07-25 00:46:45.249 4478/F9057] Remapping "/FTL-dns-cache" from 2134016 to 2138112
[2020-07-25 00:51:51.024 6126/F9057] Resizing "/FTL-strings" from 3227648 to 3231744
[2020-07-25 00:51:51.025 6126/F9057] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[2020-07-25 00:51:51.025 6126/F9057] ---------------------------->  FTL crashed!  <----------------------------
[2020-07-25 00:51:51.025 6126/F9057] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[2020-07-25 00:51:51.025 6126/F9057] Please report a bug at https://github.com/pi-hole/FTL/issues
[2020-07-25 00:51:51.025 6126/F9057] and include in your report already the following details:
[2020-07-25 00:51:51.025 6126/F9057] FTL has been running for 740098 seconds
[2020-07-25 00:51:51.025 6126/F9057] FTL branch: master
[2020-07-25 00:51:51.025 6126/F9057] FTL version: v5.1
[2020-07-25 00:51:51.025 6126/F9057] FTL commit: b522646
[2020-07-25 00:51:51.025 6126/F9057] FTL date: 2020-07-15 22:26:33 +0100
[2020-07-25 00:51:51.025 6126/F9057] FTL user: started as pihole, ended as pihole
[2020-07-25 00:51:51.025 6126/F9057] Compiled for x86_64 (compiled on CI) using gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
[2020-07-25 00:51:51.025 6126/F9057] Process details: MID: 9057
[2020-07-25 00:51:51.025 6126/F9057]                  PID: 6126
[2020-07-25 00:51:51.025 6126/F9057]                  TID: 6126
[2020-07-25 00:51:51.025 6126/F9057]                  Name: pihole-FTL
[2020-07-25 00:51:51.025 6126/F9057] Received signal: Segmentation fault
[2020-07-25 00:51:51.025 6126/F9057]      at address: 0x7fc36433cfe0
[2020-07-25 00:51:51.025 6126/F9057]      with code:  SEGV_MAPERR (Address not mapped to object)
[2020-07-25 00:51:51.026 6126/F9057] Backtrace:
[2020-07-25 00:51:51.026 6126/F9057] B[0000]: /usr/bin/pihole-FTL(+0x342af) [0x562ac48ba2af]
[2020-07-25 00:51:51.040 6126/F9057] L[0000]: /root/project/src/signals.c:119
[2020-07-25 00:51:51.040 6126/F9057] B[0001]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x12730) [0x7fc367f89730]
[2020-07-25 00:51:51.040 6126/F9057] B[0002]: /lib/x86_64-linux-gnu/libc.so.6(+0x15c1a7) [0x7fc367f121a7]
[2020-07-25 00:51:51.040 6126/F9057] B[0003]: /usr/bin/pihole-FTL(+0xe487d) [0x562ac496a87d]
[2020-07-25 00:51:51.085 6126/F9057] L[0003]: /root/project/src/database/sqlite3.c:76531
[2020-07-25 00:51:51.089 6126/F9057] B[0004]: /usr/bin/pihole-FTL(+0xe4c8a) [0x562ac496ac8a]
[2020-07-25 00:51:51.124 6126/F9057] L[0004]: /root/project/src/database/sqlite3.c:83904
[2020-07-25 00:51:51.127 6126/F9057] B[0005]: /usr/bin/pihole-FTL(in_whitelist+0xb8) [0x562ac48c2968]
[2020-07-25 00:51:51.133 6126/F9057] L[0005]: /root/project/src/database/gravity-db.c:704
[2020-07-25 00:51:51.133 6126/F9057] B[0006]: /usr/bin/pihole-FTL(+0x2af89) [0x562ac48b0f89]
[2020-07-25 00:51:51.140 6126/F9057] L[0006]: /root/project/src/dnsmasq_interface.c:261
[2020-07-25 00:51:51.140 6126/F9057] B[0007]: /usr/bin/pihole-FTL(_FTL_new_query+0x69b) [0x562ac48b248b]
[2020-07-25 00:51:51.146 6126/F9057] L[0007]: /root/project/src/dnsmasq_interface.c:435
[2020-07-25 00:51:51.147 6126/F9057] B[0008]: /usr/bin/pihole-FTL(tcp_request+0x758) [0x562ac48e24b8]
[2020-07-25 00:51:51.153 6126/F9057] L[0008]: /root/project/src/dnsmasq/forward.c:1945
[2020-07-25 00:51:51.153 6126/F9057] B[0009]: /usr/bin/pihole-FTL(+0x51d49) [0x562ac48d7d49]
[2020-07-25 00:51:51.160 6126/F9057] L[0009]: /root/project/src/dnsmasq/dnsmasq.c:1967
[2020-07-25 00:51:51.160 6126/F9057] B[0010]: /usr/bin/pihole-FTL(main_dnsmasq+0x129c) [0x562ac48d9cdc]
[2020-07-25 00:51:51.167 6126/F9057] L[0010]: /root/project/src/dnsmasq/dnsmasq.c:1206
[2020-07-25 00:51:51.167 6126/F9057] B[0011]: /usr/bin/pihole-FTL(main+0xdc) [0x562ac48ad25c]
[2020-07-25 00:51:51.173 6126/F9057] L[0011]: /root/project/src/main.c:97
[2020-07-25 00:51:51.173 6126/F9057] B[0012]: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xeb) [0x7fc367dda09b]
[2020-07-25 00:51:51.173 6126/F9057] B[0013]: /usr/bin/pihole-FTL(_start+0x2a) [0x562ac48ad3da]
[2020-07-25 00:51:51.180 6126/F9057] L[0013]: ??:?
[2020-07-25 00:51:51.181 6126/F9057] ------ Listing content of directory /dev/shm ------
[2020-07-25 00:51:51.181 6126/F9057] File Mode User:Group  Filesize Filename
[2020-07-25 00:51:51.181 6126/F9057] rwxrwxrwx root:root 260 .
[2020-07-25 00:51:51.181 6126/F9057] rwxr-xr-x root:root 3K ..
[2020-07-25 00:51:51.181 6126/F9057] rw------- pihole:pihole 4K FTL-per-client-regex
[2020-07-25 00:51:51.181 6126/F9057] rw------- pihole:pihole 2M FTL-dns-cache
[2020-07-25 00:51:51.181 6126/F9057] rw------- pihole:pihole 45K FTL-overTime
[2020-07-25 00:51:51.181 6126/F9057] rw------- pihole:pihole 3M FTL-queries
[2020-07-25 00:51:51.181 6126/F9057] rw------- pihole:pihole 4K FTL-upstreams
[2020-07-25 00:51:51.181 6126/F9057] rw------- pihole:pihole 332K FTL-clients
[2020-07-25 00:51:51.181 6126/F9057] rw------- pihole:pihole 3M FTL-domains
[2020-07-25 00:51:51.181 6126/F9057] rw------- pihole:pihole 3M FTL-strings
[2020-07-25 00:51:51.181 6126/F9057] rw------- pihole:pihole 12 FTL-settings
[2020-07-25 00:51:51.181 6126/F9057] rw------- pihole:pihole 144 FTL-counters
[2020-07-25 00:51:51.181 6126/F9057] rw------- pihole:pihole 48 FTL-lock
[2020-07-25 00:51:51.181 6126/F9057] ---------------------------------------------------
[2020-07-25 00:51:51.181 6126/F9057] Thank you for helping us to improve our FTL engine!
[2020-07-25 00:51:51.181 6126/F9057] Asking parent pihole-FTL (PID 9057) to shut down
[2020-07-25 00:51:51.181 6126/F9057] FTL fork terminated!
[2020-07-25 00:51:51.182 9057/T9058] Received: Real-time signal 2 (36 -> 2)
[2020-07-25 00:51:51.182 9057/T9058] IPv4 telnet error: Interrupted system call (4)
[2020-07-25 00:51:51.182 9057M] Shutting down...
[2020-07-25 00:51:51.182 9057M] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[2020-07-25 00:51:51.182 9057M] ---------------------------->  FTL crashed!  <----------------------------
[2020-07-25 00:51:51.182 9057M] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[2020-07-25 00:51:51.182 9057M] Please report a bug at https://github.com/pi-hole/FTL/issues
[2020-07-25 00:51:51.182 9057M] and include in your report already the following details:
[2020-07-25 00:51:51.182 9057M] FTL has been running for 740098 seconds
[2020-07-25 00:51:51.182 9057M] FTL branch: master
[2020-07-25 00:51:51.182 9057M] FTL version: v5.1
[2020-07-25 00:51:51.182 9057M] FTL commit: b522646
[2020-07-25 00:51:51.182 9057M] FTL date: 2020-07-15 22:26:33 +0100
[2020-07-25 00:51:51.182 9057M] FTL user: started as pihole, ended as pihole
[2020-07-25 00:51:51.182 9057M] Compiled for x86_64 (compiled on CI) using gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
[2020-07-25 00:51:51.182 9057M] Process details: MID: 9057
[2020-07-25 00:51:51.183 9057M]                  PID: 9057
[2020-07-25 00:51:51.183 9057M]                  TID: 9057
[2020-07-25 00:51:51.183 9057M]                  Name: pihole-FTL
[2020-07-25 00:51:51.183 9057M] Received signal: Segmentation fault
[2020-07-25 00:51:51.183 9057M]      at address: 0x7fc3667759d0
[2020-07-25 00:51:51.183 9057M]      with code:  SEGV_MAPERR (Address not mapped to object)
[2020-07-25 00:51:51.183 9057M] Backtrace:
[2020-07-25 00:51:51.183 9057M] B[0000]: /usr/bin/pihole-FTL(+0x342af) [0x562ac48ba2af]
[2020-07-25 00:51:51.183 9057/T6127] Remapping "/FTL-strings" from 3227648 to 3231744
[2020-07-25 00:51:51.200 9057M] L[0000]: /root/project/src/signals.c:119
[2020-07-25 00:51:51.201 9057M] B[0001]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x12730) [0x7fc367f89730]
[2020-07-25 00:51:51.201 9057M] B[0002]: /lib/x86_64-linux-gnu/libpthread.so.0(pthread_cancel+0x9) [0x7fc367f866c9]
[2020-07-25 00:51:51.201 9057M] B[0003]: /usr/bin/pihole-FTL(main+0x110) [0x562ac48ad290]
[2020-07-25 00:51:51.207 9057M] L[0003]: /root/project/src/main.c:105
[2020-07-25 00:51:51.207 9057M] B[0004]: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xeb) [0x7fc367dda09b]
[2020-07-25 00:51:51.207 9057M] B[0005]: /usr/bin/pihole-FTL(_start+0x2a) [0x562ac48ad3da]
[2020-07-25 00:51:51.213 9057M] L[0005]: ??:?
[2020-07-25 00:51:51.214 9057M] ------ Listing content of directory /dev/shm ------
[2020-07-25 00:51:51.214 9057M] File Mode User:Group  Filesize Filename
[2020-07-25 00:51:51.214 9057M] rwxrwxrwx root:root 260 .
[2020-07-25 00:51:51.214 9057M] rwxr-xr-x root:root 3K ..
[2020-07-25 00:51:51.214 9057M] rw------- pihole:pihole 4K FTL-per-client-regex
[2020-07-25 00:51:51.214 9057M] rw------- pihole:pihole 2M FTL-dns-cache
[2020-07-25 00:51:51.214 9057M] rw------- pihole:pihole 45K FTL-overTime
[2020-07-25 00:51:51.214 9057M] rw------- pihole:pihole 3M FTL-queries
[2020-07-25 00:51:51.214 9057M] rw------- pihole:pihole 4K FTL-upstreams
[2020-07-25 00:51:51.214 9057M] rw------- pihole:pihole 332K FTL-clients
[2020-07-25 00:51:51.214 9057M] rw------- pihole:pihole 3M FTL-domains
[2020-07-25 00:51:51.214 9057M] rw------- pihole:pihole 3M FTL-strings
[2020-07-25 00:51:51.215 9057M] rw------- pihole:pihole 12 FTL-settings
[2020-07-25 00:51:51.215 9057M] rw------- pihole:pihole 144 FTL-counters
[2020-07-25 00:51:51.215 9057M] rw------- pihole:pihole 48 FTL-lock
[2020-07-25 00:51:51.215 9057M] ---------------------------------------------------
[2020-07-25 00:51:51.215 9057M] Thank you for helping us to improve our FTL engine!
[2020-07-25 00:51:51.215 9057M] FTL terminated!
```

**Comment by DL6ER** [comment]
This should be fixed by the release of Pi-hole FTL 5.2. Feel free to re-open of create a new ticket with new information if the issue persists.
